### DISCUSSION TITLE
Additional information sources

### DISCUSSION THREAD

**Discussion by mizaki** [comictagger/comictagger#316] [Ideas]
Hi,

I believe this is on the roadmap and was wondering if there was anything that had been done so far? I'm only a dabbler in programming and have been writing a python script to use the new MangaUpdates API but obviously adding it to CT would be better. I wondered if there were any thoughts/directions decided?

Here is how I see it possibly working outside of just picking a single source:

Have the ability to order sources and then work down the enabled ones filling in missing information. E.g. Comic Vine is first and has all the data but the community rating. MangaUpdates is next in the order so use that for the rating. 

In the notes it could break out which information came from where.

The other question is the cache. Would it be better to do separate caches for each source or to have the one and make it more universal? 

If there is discussions already ongoing somewhere, please let me know.

Thanks.

**Comment by lordwelch** [comment]
Nope no other discussions have been started just mentions that it would be useful.
The primary features that would be needed to support another datasource would be
1. A search function, preferably letting us filter by publication year and other metadata, preferably would return an id referring to the series/issue
1. A way to retrieve the full metadata for given issue by either series id and issue number or issue id, note that this is currently not cached by CT as most of the time it will only be retrieved once or twice, except for when testing
1. A way to retrieve cover(s) for an issue

It would need to be refactored as it is currently very comic vine centric but a universal cache would be better in the long run
At first I think it should just be a single source at a time for simplicity, but I see it ending up as an ordered list of sources to retrieve metadata from.
Though comic vine can potentially have all of the data in practice it is usually not 100% complete though the most important being the series, title, issue and date are almost always present

**Reply by lordwelch** [reply]
Probably the main issue with supporting MangaUpdates or other manga source is that ComicTagger works almost exclusively at an issue level and MangaUpdates and MyAnimeList and almost all others that I have seen only provide a count of issues and volumes but no other information at the issue level.

**Reply by mizaki** [reply]
Yes. I did notice that. My thinking there though is some information is better than no information and that filename parsing can help. The use of these other sources is going to be mainly for manga and there being nothing in CV.

I'll try and find some examples to work with. But as an example: 
```
this cool manga chapter 1.cbz
this cool manga chapter 2.cbz
this cool manga chapter 3.cbz
```
At a minimum from filename parsing alone you can get the title and chapter/issue. The nomenclature on issue/chapter/volume seems muddy as an aside. Duplicating the series to title and using the same description over all chapters is still better than nothing imo.

**Reply by lordwelch** [reply]
Most comics do not have a title so leaving that blank would probably be best
ComicVine does have a non-zero amount of manga https://comicvine.gamespot.com/hagane-no-renkinjutsushi/4050-50835/ but yes they are definitely lacking

If I were to try putting the manga updates data into the ComicTagger metadata currently I would probably try this mapping
from https://api.mangaupdates.com/#operation/retrieveSeries note that the issue count on this list would most likely conflict with anything found on ComicVine as they appear to use the volume number as the issue number.
When using https://api.mangaupdates.com/#operation/searchSeriesPost I would probably use hit_title instead of title
```
series          = title
publisher       = either the first publisher in the array or the "Original"
year            = year
issue_count     = latest_chapter
genre           = genres (use comma's to separate)
comments        = description
volume_count    = (this may be able to be parsed out from the status field)
critical_rating = bayesian_rating (divided by 2? to keep to tho 1-5 rating?)
web_link        = url
format          = type
manga           = 'Yes'
series_group    = (possible follow the related_series list to find the main series)
credits         = authors
tags            = (categories + completed? + licensed? anime? (meaning that it has an anime))
```

**Reply by mizaki** [reply]
Looks good. I'm thinking that issue and chapter are synonymous? Volume is a collection of issues/chapters?

series = The only problem with the hit_title is it'll need to be carried over from the search. As you probably saw, the alt titles don't have language identifiers.
publisher = Original is probably best if a comma separated list isn't wanted.
volume_count = Getting the volume_count will be hit and miss as it's kind of a free form things as far as I can tell. Worth a quick shot with "vol/vol./volume". 
critical_rating = Yes, I divide by 2 on the rating currently.
format = Isn't available.
manga = I currently use type and convert "manga" to "yes".
series_group = Hadn't thought of that. Seems it would be a guess but worth a go.
tags = Currently I just take the categories but the others make sense too.

From my perspective getting anything more than a title and issue number is a win.

**Reply by lordwelch** [reply]
With chapters and volumes that seems to be what I've seen but I haven't seen much
well with hit_title ComicTagger currently always uses the search unless on the command line you specify the `--id` option

I don't believe anybody handles multiple publishers on a single comic and in reality there is only ever one but figuring out which one published the particular comic that's being tagged is something that should be handled later down the line

with volume you could try shoving it through the filename parser and see if it finds it well enough

Let me know if need anything from ComicTagger and we can see what we can accomplish

**Comment by mizaki** [comment]
I think the first thing I'll try is to rename the files and add the new schema. Do a PR for that as it's been quite a few years since I've done this.

**Comment by mizaki** [comment]
I'm thinking that a source handler is the way to go, abstract away the sources to CT main. It would provide:
```
get_capabilities # Currently just "has_issues" (not a Twitter bio state) to inform if the source has issue level information. How to handle this with multiple sources, I'm thinking if one source can do it, it equals true. Then it can be broken down with the search return.
search_for_series
get_issues_for_series
get_series_info
get_issue_info
```
The sourcehandler can go off to which ever source is picked in the options and later can handle multiple sources.

Not sure if that makes sense?

**Reply by lordwelch** [reply]
How I would go about it is to move comicvinetalker.py to it's own module and import it make sure everything still works.
Then I would create an api for the object, in other words enumerate the set of functions and properties needed to initialize and use the object as is in the code base,(eg remove all of the CV references in the functions and properties in the api). Once an api is established, ensure that the code works with the minimum changes required, then simplify the usage to better suit the api created. Once that is done I would I would try to determine the best way to figure out how to pick between two sources (store the type in a variable on TaggerWindow and use it to instantiate? what would I do for cli.py?), how to get multiple sources into a list wouldn't matter just need the infrastructure there. Once that is done I would then determine how to deal with sources that don't have the issue level information needed, primarily the cover image.

**Comment by mizaki** [comment]
What are you thoughts on how, umm... plug-able(?) should sources be? I've taken some ideas from https://github.com/gdiepen/python_plugin_example

Currently I'm thinking that there is a "sourcesapi/sources" dir with modules for each source: comicvinetalker.py, mangaupdates.py inside.

The question is; should CT read these files and import any that it finds in that dir or, make it a static list programed in? It's the security I'm thinking about, loading any modules in the sources package dir could cause problems.

Hope that makes sense?

**Reply by lordwelch** [reply]
So the nice thing about plugins in python is that it doesn't really change the execution of the plugin code itself just the code to determine what plugin gets executed. So for example making the archivers (zip,rar,7zip) into plugins is like 60% done already because they all have the same api. I plan on making a plugin system inspired by how flake8 currently does their plugins, however I've been busy with other things.

The most important thing is to make it so that initializing and using a class is as consistent as possible.

You can check #195 for an example I believe I made it using the repository you linked to but changing the plugin system to be more like flake8 won't change any of the actual plugins just the __init__ of ComicArchive
